shotover-proxy/src/frame/cassandra.rs

```diff
@@ -476,38 +505,292 @@ impl CassandraOperation {
-            Err(err) => {
-                tracing::error!("Failed to parse CQL for frame {:?}\nError: {:?}", sql, err);
-                CQL::FailedToParse(sql)
+            CassandraStatement::Update(update) => {
+                for assignment_element in &update.assignments {
+                    if let Operand::Param(_) = &assignment_element.value {
+                        return true;
+                    }
+                    if let Some(assignment_operator) = &assignment_element.operator {
+                        match assignment_operator {
+                            AssignmentOperator::Plus(operand) => {
+                                if let Operand::Param(_) = operand {
+                                    return true;
+                                }
+                            }
+                            AssignmentOperator::Minus(operand) => {
+                                if let Operand::Param(_) = operand {
+                                    return true;
+                                }
+                            }
+                        }
+                    }
+                }
+                if CQLStatement::has_params_in_relation_elements(&update.where_clause) {
+                    return true;
+                }
+                if CQLStatement::has_params_in_relation_elements(&update.if_clause) {
+                    return true;
+                }
             }
+            _ => {}
+        }
+        false
+    }
+}
+
+impl Display for CQLStatement {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.statement.fmt(f)
+    }
+}
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct CQL {
+    pub statements: Vec<CQLStatement>,
+}
+
+impl CQL {
+    fn to_query_string(&self) -> String {
+        self.statements
+            .iter()
+            .map(|c| c.statement.to_string())
+            .join("; ")
+    }
+
+    /// the CassandraAST handles multiple queries in a string separated by semi-colons: `;` however
+    /// CQL only stores one query so this method only returns the first one if there are multiples.
+    pub fn parse_from_string(cql_query_str: &str) -> Self {
+        debug!("parse_from_string: {}", cql_query_str);
+        let ast = CassandraAST::new(cql_query_str);
+
+        CQL {
+            statements: ast
+                .statements
+                .iter()
+                .map(|(_, stmt)| match (ast.has_error(), stmt) {
+                    (true, CassandraStatement::Unknown(_)) => CQLStatement {
+                        statement: stmt.clone(),
+                    },
+                    (true, _) => CQLStatement {
+                        statement: CassandraStatement::Unknown(stmt.to_string()),
+                    },
+                    (false, _) => CQLStatement {
+                        statement: stmt.clone(),
+                    },
+                })
+                .collect(),
+        }
+    }
```

Comment:
There are two problems here:
1. stmt.to_string() is not correct as that will generate a string from a partially parsed cql statement, instead we need to take the string from the original cql_query_str value.
2. cloning the entire statement is really expensive, we need to use into_iter instead of iter to prevent that

So I suggest the following implementation:
```rust
    pub fn parse_from_string(cql_query_str: &str) -> Self {
        debug!("parse_from_string: {}", cql_query_str);
        let ast = CassandraAST::new(cql_query_str);

        let error = ast.has_error();
        let statements = ast
            .statements
            .into_iter()
            .map(|(_, statement)| CQLStatement {
                statement: match (error, statement) {
                    (true, statement @ CassandraStatement::Unknown(_)) => statement,
                    (true, statement) => CassandraStatement::Unknown(cql_query_str.to_string()),
                    (false, statement) => statement,
                },
            })
            .collect();
        CQL { statements }
    }
``

Reply:
Oh, thats much better than what I had!